src/coreclr/md/enc/sha256.cpp

```diff
@@ -0,0 +1,141 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+//*****************************************************************************
+// sha256.cpp
+//
+
+//
+// contains implementation of sha256 hash algorithm
+//
+//*****************************************************************************
+
+#ifdef _WIN32
+HRESULT Sha256Hash(BYTE* pSrc, DWORD srcSize, BYTE* pDst, DWORD dstSize)
+{
+    NTSTATUS status;
+
+    BCRYPT_ALG_HANDLE   algHandle = NULL;
+    BCRYPT_HASH_HANDLE  hashHandle = NULL;
+
+    BYTE    hash[32]; // 256 bits
+    DWORD   hashLength = 0;
+    DWORD   resultLength = 0;
+    status = BCryptOpenAlgorithmProvider(&algHandle, BCRYPT_SHA256_ALGORITHM, NULL, BCRYPT_HASH_REUSABLE_FLAG);
+    if(!NT_SUCCESS(status))
+    {
+        goto cleanup;
+    }
+    status = BCryptGetProperty(algHandle, BCRYPT_HASH_LENGTH, (PBYTE)&hashLength, sizeof(hashLength), &resultLength, 0);
+    if(!NT_SUCCESS(status))
+    {
+        goto cleanup;
+    }
+    if (hashLength != 32)
+    {
+        status = STATUS_NO_MEMORY;
+        goto cleanup;
+    }
+    status = BCryptCreateHash(algHandle, &hashHandle, NULL, 0, NULL, 0, 0);
+    if(!NT_SUCCESS(status))
+    {
+        goto cleanup;
+    }
+
+    status = BCryptHashData(hashHandle, pSrc, srcSize, 0);
+    if(!NT_SUCCESS(status))
+    {
+        goto cleanup;
+    }
+
+    status = BCryptFinishHash(hashHandle, hash, hashLength, 0);
+    if(!NT_SUCCESS(status))
+    {
+        goto cleanup;
+    }
+
+    memcpy(pDst, hash, min(hashLength, dstSize));
```

Comment:
Why do we use `min` here? Is there a real scenario where `dstSize` is not big enough to hold the whole SHA-256 and it needs to be truncated?

Reply:
In #109091, it's convenient to be able to get a truncated hash that can be used as a checksum ([example](https://github.com/dotnet/runtime/blob/6780404c7a0ced64670ac2d0d2a0929b1145ad37/src/coreclr/ilasm/writer.cpp#L1707)). We could move the temp buffer logic to the caller, and require that `pDst` in `Sha256Hash` is big enough for a whole SHA-256 digest, if you'd prefer.

Reply:
> We could move the temp buffer logic to the caller

I think it would make more sense.